platform/engineering/request-for-change/2019-09-30-cms-integration.md

```diff
@@ -0,0 +1,182 @@
+send a Slack notification with a link to the build. If Jenkins was
+building the production branch, the build will also fail, to prevent
+broken links in production.
+
+## Motivation
+_Why do we want to change the current implementation? What problem(s)
+does the change solve?_
+
+During the Jenkins builds for branches, the build script fetches the
+latest content from `vagov-content` and the CMS. This means that **two
+subsequent builds on a branch with no changes to the code can have two
+different results.** In fact, this happens _very_ frequently, because
+the changed content in Drupal often introduces broken links.
+
+The goal of this RFC is to separate the content build from the
+`vets-website` Webpack build, so when `vets-website` pulls the content
+to serve up, it's already been built and validated. If this RFC is
+successful, **all the sources of build failures due to content issues
+will be isolated to the content build and will not affect
+`vets-website` builds or deploys.**
+
+## Design
+_Explain the proposed changes in enough detail so that a team member
+with working knowledge of the project could implement the change
+themselves. Bonus points if this is easily mappable to user stories or
+specs._
+
+I propose rearchitecting the `vets-website` build to separate the
+content building into a process distinct from the website.
+
+![separate build diagram](images/separated-content-build-proposal.png)
+
+**Note:** Pulling the already-built content _can still_ have two
+different results, much like the current state, but since the content
+it's pulling will have already been validated, the likelihood of a
+failure is drastically reduced.
+
+### Changes
+- `vets-website` will no longer build the content
+  - **Possible exception:** We may need to update some things in the
+    pre-built content such as references to the new cache-busted
+    filenames for JS and CSS assets
+  - **Possible exception:** It should still be responsible for building
+    the React pages, which would need to have the surrounding markup
+    come from the templates
+- There will be a new repo for the content build process
+  - This will include scripts and templates
+- There will be a new Jenkins pipeline for building the content
+  - And everything that goes with this
+- There will be a (new?) S3 bucket which holds all the built content
+  - We currently do this with the cache, so this isn't much of a
+    deviation
+- The preview server will live in the new content-building repo
+- Sitemap accessibility tests will be done during the content build,
+  not the `vets-website` build
+- The `vets-website` Jenkins build will have the following changes
+  - There will be no cache fallback
+  - There will be no content caching
+  - There will be no sitemap accessibility testing
+  
+### Serving React apps
+`vets-website` should be responsible for making the pages where the
+React apps live, but not responsible for building the header, footer,
```

Comment:
we added the date hash for debugging. maybe we just hav ethe js apps insert an html comment / adds something to the global state that outlines their version / build date info.

Reply:
>  I don't think we could get to a webpack only build if we have to pull in all of the content to inject our asset names into the markup.

I don't think we'd start with a Webpack-only `vets-website` build since it's a more drastic change, but I'd like to get there shortly after the separation. I don't know what that would look like right now, though, for the same reasons you mentioned.

> This means that the CMS would be soley responsible for managing the references to the webpack output: js / css / webfont files.

I don't think this would work. In my diagram, you can see I have the `vets-website` build pull the pre-built content. Because this content is pre-built and then saved in S3, it won't know anything about the references to the Webpack output. It also _shouldn't_ know about them. The separation of concerns goes both ways. I can definitely write more about this, but will defer that soliloquy for now. :stuck_out_tongue: 

> An alternative approach is to not create new asset names. The purpose of adding hashes to file names is for cache busting and for debugging (to a lesser extent). This means that the CMS would be soley responsible for managing the references to the webpack output: js / css / webfont files. We can cache-control header + etags to do cache busting. There's a slight performance trade off to use this too.

Adding hashes to the asset filenames aside, we'd still have to deal with adding `build/${env}/path/to/react/app/index.html` and ensuring it has the correct assets in that file. As I said above, this shouldn't happen during the CMS build, since the React apps are `vets-website`'s responsibility. Additionally, the paths to those applications live in their respective `manifest.json` files, so the CMS build would have to be aware of those, which then creates a fairly tight coupling between the two, which we're trying to avoid as much as possible.

That said, I think you're right about the cache busting. I'm not super familiar with the reasoning behind the way we do it now, but I like where you're going with that. It's probably tangential to this particular RFC, though, and can be done with or without this.

Reply:
I forgot about the environment targeting. I guess that means there will always be a need for js build to inject something into the static markup. 

I guess we can profile it and see. My initial thought is that pulling in all of the content and adding the script and link tag is going to be time consuming. it would be nice if we could just fetch the content and use it as is- it would also be nice to not have to take an output from one process and put it into another. especially if that output rarely changes e.g. our js common bundle name has probably never changed besides the cache busting hashing.

Reply:
Actually, now that I think about it, if we're not already injecting asset names into the HTML files during the Webpack step, we probably won't need to do it once we separate them. We currently don't have a separate Metalsmith plugin for this.

I haven't dug into our Webpack setup to figure out how this is handled right now. I was making some assumptions... :grimacing: 

I'll dig into this today.

Reply:
Seems like over time there will be more and more apps that will be using the platform and shared common files (js) for auth / login as an example and likely leveraging existing code that is maintained by the platform team and (css) for shared common styles is something that would need to be managed as part of the platform and not have to be part of the new app teams boundary.

Reply:
@robbiethegeek-usds yeah I think that's the direction we're moving. we have a [task](https://github.com/department-of-veterans-affairs/va.gov-team/issues/1425) to investigate the micro frontend [pattern](https://martinfowler.com/articles/micro-frontends.html) that thoughtworks is pushing. I think they do a good summary of the tradeoffs. 

I think the ideal is for the apps teams have their own deployment pipeline and use common dependencies provided by the platform (maybe through an npm module). There's a lot to work through to get to that point though including defining / enforcing core quality metrics (e.g. performance budgets).

I think that autonomy includes static content deployment but I'm not sure what that looks like. Right now this process is managing a giant graphQL query and there's a lot of `content model -> template model` transformations that don't seem very clearcut / clean (i.e. easy to separate into team owned deployment pipelines). I think separating the process from the js build is a first step to supporting the principles in that micro frontends article but there's a lot of unknowns specific to `vets-website` we need to work through.